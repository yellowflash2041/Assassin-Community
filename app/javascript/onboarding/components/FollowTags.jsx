import { h, Component } from 'preact';
import PropTypes from 'prop-types';

import Navigation from './Navigation';
import { getContentOfToken } from '../utilities';

class FollowTags extends Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
    this.handleComplete = this.handleComplete.bind(this);

    this.state = {
      allTags: [],
      selectedTags: [],
    };
  }

  componentDidMount() {
    fetch('/api/tags/onboarding')
      .then(response => response.json())
      .then(data => {
        this.setState({ allTags: data });
      });

    const csrfToken = getContentOfToken('csrf-token');
    fetch('/onboarding_update', {
      method: 'PATCH',
      headers: {
        'X-CSRF-Token': csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        user: { last_onboarding_page: 'follow tags page' },
      }),
      credentials: 'same-origin',
    });
  }

  handleClick(tag) {
    let { selectedTags } = this.state;
    if (!selectedTags.includes(tag)) {
      this.setState(prevState => ({
        selectedTags: [...prevState.selectedTags, tag],
      }));
    } else {
      selectedTags = [...selectedTags];
      const indexToRemove = selectedTags.indexOf(tag);
      selectedTags.splice(indexToRemove, 1);
      this.setState({
        selectedTags,
      });
    }
  }

  handleComplete() {
    const csrfToken = getContentOfToken('csrf-token');
    const { selectedTags } = this.state;

    Promise.all(
      selectedTags.map(tag =>
        fetch('/follows', {
          method: 'POST',
          headers: {
            'X-CSRF-Token': csrfToken,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            followable_type: 'Tag',
            followable_id: tag.id,
            verb: 'follow',
          }),
          credentials: 'same-origin',
        }),
      ),
    ).then(_ => {
      const { next } = this.props;
      next();
    });
  }

  render() {
    const { prev } = this.props;
    const { selectedTags, allTags } = this.state;
    return (
      <div className="onboarding-main">
        <div className="onboarding-content">
          <h2>Follow tags to customize your feed</h2>
          <div className="scroll">
            {allTags.map(tag => (
              <button
                type="button"
                onClick={() => this.handleClick(tag)}
                style={{
                  backgroundColor: tag.bg_color_hex,
                  color: tag.text_color_hex,
                }}
                className={
                  selectedTags.includes(tag) ? 'tag tag-selected' : 'tag'
                }
              >
                #
                {tag.name}
                <div className="onboarding-tag-follow-indicator">
                  {selectedTags.includes(tag) ? 'âœ“ following ' : ''}
                </div>
              </button>
            ))}
          </div>
        </div>
        <Navigation prev={prev} next={this.handleComplete} />
      </div>
    );
  }
}

FollowTags.propTypes = {
  prev: PropTypes.func.isRequired,
  next: PropTypes.string.isRequired,
};

export default FollowTags;
