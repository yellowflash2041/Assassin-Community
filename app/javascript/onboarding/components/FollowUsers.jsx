import { h, Component } from 'preact';
import PropTypes from 'prop-types';

import Navigation from './Navigation';
import { getContentOfToken } from '../utilities';

class FollowUsers extends Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
    this.handleComplete = this.handleComplete.bind(this);

    this.state = {
      users: [],
      selectedUsers: [],
    };
  }

  componentDidMount() {
    fetch('/api/users?state=follow_suggestions', {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      credentials: 'same-origin',
    })
      .then(response => response.json())
      .then(data => {
        this.setState({ users: data, selectedUsers: data });
      });

    const csrfToken = getContentOfToken('csrf-token');
    fetch('/onboarding_update', {
      method: 'PATCH',
      headers: {
        'X-CSRF-Token': csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        user: { last_onboarding_page: 'follow users page' },
      }),
      credentials: 'same-origin',
    });
  }

  handleComplete() {
    const csrfToken = getContentOfToken('csrf-token');
    const { selectedUsers } = this.state;
    const { next } = this.props;

    fetch('/api/follows', {
      method: 'POST',
      headers: {
        'X-CSRF-Token': csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ users: selectedUsers }),
      credentials: 'same-origin',
    });

    next();
  }

  handleSelectAll() {
    const { selectedUsers, users } = this.state;
    if (selectedUsers.length === users.length) {
      this.setState({
        selectedUsers: [],
      });
    } else {
      this.setState({
        selectedUsers: users,
      });
    }
  }

  handleClick(user) {
    let { selectedUsers } = this.state;

    if (!selectedUsers.includes(user)) {
      this.setState(prevState => ({
        selectedUsers: [...prevState.selectedUsers, user],
      }));
    } else {
      selectedUsers = [...selectedUsers];
      const indexToRemove = selectedUsers.indexOf(user);
      selectedUsers.splice(indexToRemove, 1);
      this.setState({
        selectedUsers,
      });
    }
  }

  render() {
    const { users, selectedUsers } = this.state;
    const { prev } = this.props;
    return (
      <div className="onboarding-main">
        <div className="onboarding-content">
          <h2>Ok, here are some people we picked for you</h2>
          <div className="scroll">
            <div className="select-all-button-wrapper">
              <button type="button" onClick={() => this.handleSelectAll()}>
                Select All 
                {' '}
                {selectedUsers.length === users.length ? 'âœ…' : ''}
              </button>
            </div>
            {users.map(user => (
              <button
                type="button"
                style={{
                  backgroundColor: selectedUsers.includes(user)
                    ? '#c7ffe8'
                    : 'white',
                }}
                onClick={() => this.handleClick(user)}
                className="user"
              >
                <div className="onboarding-user-follow-status">
                  {selectedUsers.includes(user) ? 'selected' : ''}
                </div>
                <img src={user.profile_image_url} alt="" />
                <span>{user.name}</span>
                <p>{user.summary}</p>
              </button>
            ))}
          </div>
        </div>
        <Navigation prev={prev} next={this.handleComplete} />
      </div>
    );
  }
}

FollowUsers.propTypes = {
  prev: PropTypes.func.isRequired,
  next: PropTypes.string.isRequired,
};

export default FollowUsers;
