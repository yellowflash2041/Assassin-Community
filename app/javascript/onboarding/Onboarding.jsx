import 'preact/devtools';
import { h, Component } from 'preact';

import IntroSlide from './components/IntroSlide';
import PersonalInfoForm from './components/PersonalInfoForm';
import EmailListTermsConditionsForm from './components/EmailListTermsConditionsForm';
import ClosingSlide from './components/ClosingSlide';
import FollowTags from './components/FollowTags';
import FollowUsers from './components/FollowUsers';
import BioForm from './components/BioForm';

// Current Onboarding Variants
// 0) Original intro slide: three explainer paragraphs, left adjusted.
// 1) Modified intro slide: Cat gif, let's get started.
// 2) Modified intro slide: Cat gif, We have a few quick questions to fill out your profile
// 3) Modified intro slide: Skull gif, The more you get involved in community, the better developer you will be.
// 4) Modified intro slide: Skull gif, You just made a great choice for your dev career
// 5) No intro slide.
// 6) Last slide challenge: Leave three constructive comments today
// 7) Last slide: Only display cue for welcome thread
// 8) Last slide: Only display cue for welcome thread, and also the leave three constructive comments challenge
// 9) Last slide: Only display cue for welcome thread, display challenge right in link body

export default class Onboarding extends Component {
  constructor(props) {
    super(props);

    const url = new URL(window.location);
    const previousLocation = url.searchParams.get('referrer');
    let variant = '0';
    if (url.searchParams.get('variant') || window.currentUser) {
      variant =
        url.searchParams.get('variant') ||
        window.currentUser.onboarding_variant_version ||
        '0';
    }

    this.nextSlide = this.nextSlide.bind(this);
    this.prevSlide = this.prevSlide.bind(this);

    let slides = [
      IntroSlide,
      EmailListTermsConditionsForm,
      BioForm,
      PersonalInfoForm,
      FollowTags,
      FollowUsers,
      ClosingSlide,
    ];

    if (variant === '5') {
      slides = [
        EmailListTermsConditionsForm,
        BioForm,
        PersonalInfoForm,
        FollowTags,
        FollowUsers,
        ClosingSlide,
      ];
    }

    this.slides = slides.map(SlideComponent => (
      <SlideComponent
        next={this.nextSlide}
        prev={this.prevSlide}
        previousLocation={previousLocation}
        variant={variant}
      />
    ));

    this.state = {
      currentSlide: 0,
    };
  }

  nextSlide() {
    const { currentSlide } = this.state;
    const nextSlide = currentSlide + 1;
    if (nextSlide < this.slides.length) {
      this.setState({
        currentSlide: nextSlide,
      });
    }
  }

  prevSlide() {
    const { currentSlide } = this.state;
    const prevSlide = currentSlide - 1;
    if (prevSlide >= 0) {
      this.setState({
        currentSlide: prevSlide,
      });
    }
  }

  render() {
    const { currentSlide } = this.state;
    return <div className="onboarding-body">{this.slides[currentSlide]}</div>;
  }
}
