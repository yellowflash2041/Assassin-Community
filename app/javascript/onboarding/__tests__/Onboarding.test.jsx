import { h } from 'preact';
import { deep } from 'preact-render-spy';
import fetch from 'jest-fetch-mock';
import { axe } from 'jest-axe';

import Onboarding from '../Onboarding';
import ProfileForm from '../components/ProfileForm';
import FollowTags from '../components/FollowTags';
import FollowUsers from '../components/FollowUsers';

global.fetch = fetch;

// Corresponds to the order of slides declared in the Onboarding component.
const slides = [
  'IntroSlide',
  'FollowTags',
  'ProfileForm',
  'FollowUsers',
  'EmailPreferencesForm',
];

function flushPromises() {
  return new Promise((resolve) => setImmediate(resolve));
}

function initializeSlides(currentSlide, userData = null, mockData = null) {
  document.body.setAttribute('data-user', userData);
  const onboardingSlides = deep(
    <Onboarding
      communityConfig={{
        communityName: 'Test',
        communityDescription: "Wouldn't you like to knoww..",
      }}
    />,
  );

  if (mockData) {
    fetch.once(mockData);
  }

  onboardingSlides.setState({ currentSlide });

  return onboardingSlides;
}

function expectStepperToRender(onboardingSlides, activeDotCount) {
  // We do not show the stepper on the IntroSlide.
  const stepperDotCount = slides.length - 1;

  expect(onboardingSlides.find('.stepper').exists()).toEqual(true);
  expect(onboardingSlides.find('.dot').length).toBe(stepperDotCount);
  expect(onboardingSlides.find('.active').length).toBe(activeDotCount);
}

describe('<Onboarding />', () => {
  beforeEach(() => {
    fetch.resetMocks();
  });

  // Use this to fetch mock response data before trying to render the `followTags` component.
  const fakeTagsResponse = JSON.stringify([
    {
      bg_color_hex: '#000000',
      id: 715,
      name: 'discuss',
      text_color_hex: '#ffffff',
    },
    {
      bg_color_hex: '#f7df1e',
      id: 6,
      name: 'javascript',
      text_color_hex: '#000000',
    },
    {
      bg_color_hex: '#2a2566',
      id: 630,
      name: 'career',
      text_color_hex: '#ffffff',
    },
  ]);

  // Use this to fetch mock response data before trying to render the `followUsers` component.
  const fakeUsersResponse = JSON.stringify([
    {
      id: 1,
      name: 'Ben Halpern',
      profile_image_url: 'apple-icon.png',
    },
    {
      id: 2,
      name: 'Krusty the Clown',
      profile_image_url: 'clown.jpg',
    },
    {
      id: 3,
      name: 'dev.to staff',
      profile_image_url: 'dev.jpg',
    },
  ]);
  const getUserData = () =>
    JSON.stringify({
      followed_tag_names: ['javascript'],
      profile_image_90: 'mock_url_link',
      name: 'firstname lastname',
      username: 'username',
    });

  describe('IntroSlide', () => {
    const slideIndex = slides.indexOf('IntroSlide');
    let onboardingSlides;
    const codeOfConductCheckEvent = {
      target: {
        value: 'checked_code_of_conduct',
        name: 'checked_code_of_conduct',
      },
    };
    const termsAndConditionsCheckEvent = {
      target: {
        value: 'checked_terms_and_conditions',
        name: 'checked_terms_and_conditions',
      },
    };

    const updateCodeOfConduct = () => {
      onboardingSlides
        .find('#checked_code_of_conduct')
        .simulate('change', codeOfConductCheckEvent);
    };
    const updateTermsAndConditions = () => {
      onboardingSlides
        .find('#checked_terms_and_conditions')
        .simulate('change', termsAndConditionsCheckEvent);
    };

    beforeEach(() => {
      onboardingSlides = initializeSlides(slideIndex, getUserData());
    });

    it('renders properly', () => {
      expect(onboardingSlides).toMatchSnapshot();
    });

    it('it does not render a stepper', () => {
      expect(onboardingSlides.find('.stepper').length).toBe(0);
    });

    it('should advance if required boxes are checked', async () => {
      fetch.once({});
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex);

      updateCodeOfConduct();
      updateTermsAndConditions();

      onboardingSlides.find('.next-button').simulate('click');

      // Fetch the fakeTagsResponse before trying to render the next slide (followTags).
      fetch.once(fakeTagsResponse);
      await flushPromises();
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex + 1);
    });

    it('should not have basic a11y violations', async () => {
      const results = await axe(onboardingSlides.toString());

      expect(results).toHaveNoViolations();
    });
  });

  describe('FollowTags', () => {
    let onboardingSlides;
    const slideIndex = slides.indexOf('FollowTags');

    beforeEach(async () => {
      onboardingSlides = initializeSlides(
        slideIndex,
        getUserData(),
        fakeTagsResponse,
      );
      await flushPromises();
    });

    it('renders properly', () => {
      expect(onboardingSlides).toMatchSnapshot();
    });

    it('renders a stepper', () => {
      expectStepperToRender(onboardingSlides, slideIndex);
    });

    it('should render three tags', async () => {
      expect(onboardingSlides.find('.onboarding-tags__item').length).toBe(3);
    });

    it('should allow a user to add a tag and advance', async () => {
      fetch.once({});
      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Skip for now',
      );

      const firstButton = onboardingSlides
        .find('.onboarding-tags__button')
        .first();

      firstButton.simulate('click');
      const followTags = onboardingSlides.find(<FollowTags />);
      expect(followTags.state('selectedTags').length).toBe(1);

      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Continue',
      );

      onboardingSlides.find('.next-button').simulate('click');
      fetch.once(fakeUsersResponse);
      await flushPromises();
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex + 1);
    });

    it('should step backward', () => {
      onboardingSlides.find('.back-button').simulate('click');
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex - 1);
    });
  });

  describe('ProfileForm', () => {
    let onboardingSlides;
    const slideIndex = slides.indexOf('ProfileForm');
    const meta = document.createElement('meta');

    meta.setAttribute('name', 'csrf-token');
    document.body.appendChild(meta);

    beforeEach(() => {
      onboardingSlides = initializeSlides(slideIndex, getUserData());
    });

    it('renders properly', () => {
      expect(onboardingSlides).toMatchSnapshot();
    });

    it('renders a stepper', () => {
      expectStepperToRender(onboardingSlides, slideIndex);
    });

    it('should allow user to fill forms and advance', async () => {
      fetch.once({});
      const summaryEvent = { target: { value: 'my bio', name: 'summary' } };
      const locationEvent = {
        target: { value: 'my location', name: 'location' },
      };
      const titleEvent = {
        target: { value: 'my title', name: 'employment_title' },
      };
      const employerEvent = {
        target: { value: 'my employer name', name: 'employer_name' },
      };

      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Skip for now',
      );

      onboardingSlides.find('textarea').simulate('change', summaryEvent);
      onboardingSlides.find('#location').simulate('change', locationEvent);
      onboardingSlides.find('#employment_title').simulate('change', titleEvent);
      onboardingSlides.find('#employer_name').simulate('change', employerEvent);
      const profileForm = onboardingSlides.find(<ProfileForm />);
      expect(profileForm.state().formValues.summary).toBe(
        summaryEvent.target.value,
      );
      expect(profileForm.state().formValues.location).toBe(
        locationEvent.target.value,
      );
      expect(profileForm.state().formValues.employment_title).toBe(
        titleEvent.target.value,
      );
      expect(profileForm.state().formValues.employer_name).toBe(
        employerEvent.target.value,
      );

      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Continue',
      );

      profileForm.find('.next-button').simulate('click');
      fetch.once(fakeTagsResponse);
      await flushPromises();
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex + 1);
    });

    it('should step backward', () => {
      // Fetch the fakeTagsResponse before trying to render the previous slide (followTags).
      fetch.once(fakeTagsResponse);
      onboardingSlides.find('.back-button').simulate('click');
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex - 1);
    });
  });

  describe('FollowUsers', () => {
    let onboardingSlides;
    const slideIndex = slides.indexOf('FollowUsers');

    beforeEach(async () => {
      onboardingSlides = initializeSlides(
        slideIndex,
        getUserData(),
        fakeUsersResponse,
      );
      await flushPromises();
    });

    it('renders properly', () => {
      expect(onboardingSlides).toMatchSnapshot();
    });

    it('renders a stepper', () => {
      expectStepperToRender(onboardingSlides, slideIndex);
    });

    it('should render three users', async () => {
      expect(onboardingSlides.find('.user').length).toBe(3);
    });

    it('should allow a user to select and advance', async () => {
      fetch.once({});

      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Skip for now',
      );

      onboardingSlides.find('.user').first().simulate('click');
      expect(onboardingSlides.find('p').last().text()).toBe(
        "You're following 1 person",
      );
      onboardingSlides.find('.user').last().simulate('click');
      expect(onboardingSlides.find('p').last().text()).toBe(
        "You're following 2 people",
      );

      expect(onboardingSlides.find('.next-button').text()).toContain(
        'Continue',
      );

      const followUsers = onboardingSlides.find(<FollowUsers />);
      expect(followUsers.state('selectedUsers').length).toBe(2);
      onboardingSlides.find('.next-button').simulate('click');
      await flushPromises();
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex + 1);
    });

    it('should have a functioning select-all toggle', async () => {
      fetch.once({});

      expect(onboardingSlides.find('button').last().text()).toBe(
        'Select all 3 people',
      );
      onboardingSlides.find('button').last().simulate('click');
      expect(onboardingSlides.find('button').last().text()).toBe(
        'Deselect all',
      );
      const followUsers = onboardingSlides.find(<FollowUsers />);
      expect(followUsers.state('selectedUsers').length).toBe(3);
    });

    it('should step backward', async () => {
      onboardingSlides.find('.back-button').simulate('click');
      await flushPromises();
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex - 1);
    });
  });

  describe('EmailPreferencesForm', () => {
    let onboardingSlides;
    const { location } = window;
    const slideIndex = slides.indexOf('EmailPreferencesForm');

    beforeEach(() => {
      onboardingSlides = initializeSlides(slideIndex, getUserData());
    });

    afterEach(() => {
      window.location = location;
    });

    it('renders properly', () => {
      expect(onboardingSlides).toMatchSnapshot();
    });

    it('renders a stepper', () => {
      expectStepperToRender(onboardingSlides, slideIndex);
    });

    it('should redirect user when finished', async () => {
      fetch.once({});

      // Setup: Enable window.location to be writable.
      const url = 'https://dummy.com/onboarding';
      Object.defineProperty(window, 'location', {
        value: { href: url },
        writable: true,
      });

      expect(window.location.href).toBe(url);
      onboardingSlides.find('.next-button').simulate('click');
      await flushPromises();
      // No longer advance slide.
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex);
      expect(window.location.href).toBe('/');
    });

    it('should step backward', () => {
      onboardingSlides.find('.back-button').simulate('click');
      expect(onboardingSlides.state().currentSlide).toBe(slideIndex - 1);
    });
  });
});
