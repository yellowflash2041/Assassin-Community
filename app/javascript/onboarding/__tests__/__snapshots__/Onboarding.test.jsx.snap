// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Onboarding /> EmailPreferencesForm renders properly 1`] = `
preact-render-spy (1 nodes)
-------
<main class="onboarding-body">
  <div class="onboarding-main crayons-modal">
    <div class="crayons-modal__box">
      <nav class="onboarding-navigation">
        <div class="navigation-content">
          <div class="back-button-container">
            <button
              onClick={[Function bound prevSlide]}
              class="back-button"
              type="button"
            >
              <svg
                width="24"
                height="24"
                fill="none"
                class="crayons-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414L7.828 11z"></path>
              </svg>
            </button>
          </div>
          <div class="stepper">
            <span class="dot active"></span>
            <span class="dot active"></span>
            <span class="dot active"></span>
            <span class="dot active"></span>
          </div>
          <button
            disabled={false}
            onClick={[Function bound onSubmit]}
            class="next-button"
            type="button"
          >
            Finish
          </button>
        </div>
      </nav>
      <div class="onboarding-content terms-and-conditions-wrapper">
        <header class="onboarding-content-header">
          <h1 class="title">Almost there!</h1>
          <h2 class="subtitle">
            Review your email preferences before we continue.
          </h2>
        </header>
        <form>
          <fieldset>
            <legend>Email preferences</legend>
            <ul>
              <li class="checkbox-item">
                <label htmlFor="email_newsletter">
                  <input
                    type="checkbox"
                    id="email_newsletter"
                    name="email_newsletter"
                    checked={true}
                    onChange={[Function bound handleChange]}
                   />
                  I want to receive weekly newsletter emails.
                </label>
              </li>
              <li class="checkbox-item">
                <label htmlFor="email_digest_periodic">
                  <input
                    type="checkbox"
                    id="email_digest_periodic"
                    name="email_digest_periodic"
                    checked={true}
                    onChange={[Function bound handleChange]}
                   />
                  I want to receive a periodic digest of top posts from my tags.
                </label>
              </li>
            </ul>
          </fieldset>
        </form>
      </div>
    </div>
  </div>
</main>

`;

exports[`<Onboarding /> FollowTags renders properly 1`] = `
preact-render-spy (1 nodes)
-------
<main class="onboarding-body">
  <div class="onboarding-main crayons-modal">
    <div class="crayons-modal__box overflow-auto">
      <nav class="onboarding-navigation">
        <div class="navigation-content">
          <div class="back-button-container">
            <button
              onClick={[Function bound prevSlide]}
              class="back-button"
              type="button"
            >
              <svg
                width="24"
                height="24"
                fill="none"
                class="crayons-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414L7.828 11z"></path>
              </svg>
            </button>
          </div>
          <div class="stepper">
            <span class="dot active"></span>
            <span class="dot "></span>
            <span class="dot "></span>
            <span class="dot "></span>
          </div>
          <button
            disabled={false}
            onClick={[Function bound handleComplete]}
            class="next-button skip-for-now"
            type="button"
          >
            Skip for now
          </button>
        </div>
      </nav>
      <div class="onboarding-content toggle-bottom">
        <header class="onboarding-content-header">
          <h1 class="title">What are you interested in?</h1>
          <h2 class="subtitle">Follow tags to customize your feed</h2>
        </header>
        <div class="onboarding-modal-scroll-container">
          <div class="onboarding-tags">
            <div
              class="onboarding-tags__item false"
              style="box-shadow: inset 0 0 0 2px #000000; color: ;"
            >
              <div class="onboarding-tags__item__inner">
                #discuss
                <button
                  type="button"
                  onClick={[Function onClick]}
                  class="onboarding-tags__button false"
                  style="background-color: #000000; color: #ffffff;"
                >
                  Follow
                </button>
              </div>
            </div>
            <div
              class="onboarding-tags__item false"
              style="box-shadow: inset 0 0 0 2px #f7df1e; color: ;"
            >
              <div class="onboarding-tags__item__inner">
                #javascript
                <button
                  type="button"
                  onClick={[Function onClick]}
                  class="onboarding-tags__button false"
                  style="background-color: #f7df1e; color: #000000;"
                >
                  Follow
                </button>
              </div>
            </div>
            <div
              class="onboarding-tags__item false"
              style="box-shadow: inset 0 0 0 2px #2a2566; color: ;"
            >
              <div class="onboarding-tags__item__inner">
                #career
                <button
                  type="button"
                  onClick={[Function onClick]}
                  class="onboarding-tags__button false"
                  style="background-color: #2a2566; color: #ffffff;"
                >
                  Follow
                </button>
              </div>
            </div>
          </div>
        </div>
        <div class="onboarding-selection-status">
          <div class="selection-status-content">
            <p class="follow-count-message">0 tags selected</p>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>

`;

exports[`<Onboarding /> FollowUsers renders properly 1`] = `
preact-render-spy (1 nodes)
-------
<main class="onboarding-body">
  <div class="onboarding-main crayons-modal">
    <div class="crayons-modal__box overflow-auto">
      <nav class="onboarding-navigation">
        <div class="navigation-content">
          <div class="back-button-container">
            <button
              onClick={[Function bound prevSlide]}
              class="back-button"
              type="button"
            >
              <svg
                width="24"
                height="24"
                fill="none"
                class="crayons-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414L7.828 11z"></path>
              </svg>
            </button>
          </div>
          <div class="stepper">
            <span class="dot active"></span>
            <span class="dot active"></span>
            <span class="dot active"></span>
            <span class="dot "></span>
          </div>
          <button
            disabled={false}
            onClick={[Function bound handleComplete]}
            class="next-button skip-for-now"
            type="button"
          >
            Skip for now
          </button>
        </div>
      </nav>
      <div class="onboarding-content toggle-bottom">
        <header class="onboarding-content-header">
          <h1 class="title">Suggested people to follow</h1>
          <h2 class="subtitle">Let's review a few things first</h2>
        </header>
        <div class="onboarding-modal-scroll-container">
          <button
            type="button"
            onClick={[Function onClick]}
            onKeyDown={[Function onKeyDown]}
            class="user content-row unselected"
          >
            <figure class="user-avatar-container">
              <img
                class="user-avatar"
                src="apple-icon.png"
                alt="profile"
               />
            </figure>
            <div class="user-info">
              <h4 class="user-name">Ben Halpern</h4>
              <p class="user-summary"></p>
            </div>
            <button
              type="button"
              class="user-following-status"
            >
              Follow
            </button>
          </button>
          <button
            type="button"
            onClick={[Function onClick]}
            onKeyDown={[Function onKeyDown]}
            class="user content-row unselected"
          >
            <figure class="user-avatar-container">
              <img
                class="user-avatar"
                src="clown.jpg"
                alt="profile"
               />
            </figure>
            <div class="user-info">
              <h4 class="user-name">Krusty the Clown</h4>
              <p class="user-summary"></p>
            </div>
            <button
              type="button"
              class="user-following-status"
            >
              Follow
            </button>
          </button>
          <button
            type="button"
            onClick={[Function onClick]}
            onKeyDown={[Function onKeyDown]}
            class="user content-row unselected"
          >
            <figure class="user-avatar-container">
              <img
                class="user-avatar"
                src="dev.jpg"
                alt="profile"
               />
            </figure>
            <div class="user-info">
              <h4 class="user-name">dev.to staff</h4>
              <p class="user-summary"></p>
            </div>
            <button
              type="button"
              class="user-following-status"
            >
              Follow
            </button>
          </button>
        </div>
      </div>
      <div class="onboarding-selection-status">
        <div class="selection-status-content">
          <p class="follow-count-message">You're not following anyone</p>
          <button
            type="button"
            onClick={[Function onClick]}
          >
            Select all 3 people
          </button>
        </div>
      </div>
    </div>
  </div>
</main>

`;

exports[`<Onboarding /> IntroSlide renders properly 1`] = `
preact-render-spy (1 nodes)
-------
<main class="onboarding-body">
  <div class="onboarding-main introduction crayons-modal crayons-modal--m">
    <div class="crayons-modal__box overflow-auto">
      <div class="onboarding-content">
        <figure>
          <img
            src="/assets/purple-dev-logo.png"
            class="sticker-logo"
            alt="DEV"
           />
        </figure>
        <h1 class="introduction-title">firstname lastname— welcome to Test!</h1>
        <h2 class="introduction-subtitle">Wouldn't you like to knoww..</h2>
      </div>
      <div class="checkbox-form-wrapper">
        <form class="checkbox-form">
          <fieldset>
            <ul>
              <li class="checkbox-item">
                <label htmlFor="checked_code_of_conduct">
                  <input
                    type="checkbox"
                    id="checked_code_of_conduct"
                    name="checked_code_of_conduct"
                    checked={false}
                    onChange={[Function bound handleChange]}
                   />
                  You agree to uphold our 
                  <a
                    href="/code-of-conduct"
                    data-no-instant={true}
                    onClick={[Function onClick]}
                  >
                    Code of Conduct
                  </a>
                  .
                </label>
              </li>
              <li class="checkbox-item">
                <label htmlFor="checked_terms_and_conditions">
                  <input
                    type="checkbox"
                    id="checked_terms_and_conditions"
                    name="checked_terms_and_conditions"
                    checked={false}
                    onChange={[Function bound handleChange]}
                   />
                  You agree to our 
                  <a
                    href="/terms"
                    data-no-instant={true}
                    onClick={[Function onClick]}
                  >
                    Terms and Conditions
                  </a>
                  .
                </label>
              </li>
            </ul>
          </fieldset>
        </form>
        <nav class="onboarding-navigation intro-slide">
          <div class="navigation-content intro-slide">
            <button
              disabled={true}
              onClick={[Function bound onSubmit]}
              class="next-button"
              type="button"
            >
              Continue
            </button>
          </div>
        </nav>
      </div>
    </div>
  </div>
</main>

`;

exports[`<Onboarding /> ProfileForm renders properly 1`] = `
preact-render-spy (1 nodes)
-------
<main class="onboarding-body">
  <div class="onboarding-main crayons-modal">
    <div class="crayons-modal__box">
      <nav class="onboarding-navigation">
        <div class="navigation-content">
          <div class="back-button-container">
            <button
              onClick={[Function bound prevSlide]}
              class="back-button"
              type="button"
            >
              <svg
                width="24"
                height="24"
                fill="none"
                class="crayons-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path d="M7.828 11H20v2H7.828l5.364 5.364-1.414 1.414L4 12l7.778-7.778 1.414 1.414L7.828 11z"></path>
              </svg>
            </button>
          </div>
          <div class="stepper">
            <span class="dot active"></span>
            <span class="dot active"></span>
            <span class="dot "></span>
            <span class="dot "></span>
          </div>
          <button
            disabled={false}
            onClick={[Function bound onSubmit]}
            class="next-button skip-for-now"
            type="button"
          >
            Skip for now
          </button>
        </div>
      </nav>
      <div class="onboarding-content about">
        <header class="onboarding-content-header">
          <h1 class="title">Build your profile</h1>
          <h2 class="subtitle">
            Tell us a little bit about yourself — this is how others will see you on Test. You’ll always be able to edit this later in your Settings.
          </h2>
        </header>
        <div class="current-user-info">
          <figure class="current-user-avatar-container">
            <img
              class="current-user-avatar"
              alt="profile"
              src="mock_url_link"
             />
          </figure>
          <h3>firstname lastname</h3>
          <p>username</p>
        </div>
        <form>
          <label htmlFor="summary">
            Bio
            <textarea
              name="summary"
              id="summary"
              placeholder="Tell us about yourself"
              onChange={[Function bound handleChange]}
              maxLength="120"
            >
            </textarea>
          </label>
          <label htmlFor="location">
            Where are you located?
            <input
              type="text"
              name="location"
              id="location"
              placeholder="e.g. New York, NY"
              onChange={[Function bound handleChange]}
              maxLength="60"
             />
          </label>
          <label htmlFor="employment_title">
            What is your title?
            <input
              type="text"
              name="employment_title"
              id="employment_title"
              placeholder="e.g. Software Engineer"
              onChange={[Function bound handleChange]}
              maxLength="60"
             />
          </label>
          <label htmlFor="employer_name">
            Where do you work?
            <input
              type="text"
              name="employer_name"
              id="employer_name"
              placeholder="e.g. Company name, self-employed, etc."
              onChange={[Function bound handleChange]}
              maxLength="60"
              class="onboarding-form-input--last"
             />
          </label>
        </form>
      </div>
    </div>
  </div>
</main>

`;
